src/audio/buffer.c

```diff
@@ -176,8 +176,6 @@ void buffer_free(struct comp_buffer *buffer)
 	/* In case some listeners didn't unregister from buffer's callbacks */
 	notifier_unregister_all(NULL, buffer);
 
-	list_item_del(&buffer->source_list);
-	list_item_del(&buffer->sink_list);
```

Comment:
interesting. `list_item_del()` in SOF also does `list_init()`, so, you can delete list items as often as you like. And you can use `list_is_empty()` on list items. So this shouldn't hurt the run-time, although this might be redundant. So firstly I don't understand why valgrind complains, and secondly - are we sure `buffer_free()` is only ever called on buffers, already deleted from the lists? That *should* be the case... Maybe add an assert here?

Reply:
there is pipeline_connect() and pipeline_disconnect() for the buffer list manipulation. This was an extra delete that could also impact with the ordering that the driver frees components (hence the valgrind warning and fuzzer warning below).
```
Project: sound-open-firmware
Fuzzing Engine: libFuzzer
Fuzz Target: fuzz_ipc
Job Type: libfuzzer_ubsan_sound-open-firmware
Platform Id: linux

Crash Type: Null-dereference
Crash Address: 
Crash State:
  buffer_free
  ipc_buffer_free
  ipc_glb_tplg_free
  
Sanitizer: undefined (UBSAN)
```